modules/infra/datapath/trace.c

```diff
@@ -0,0 +1,149 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2024 Robin Jarry
+
+#include "gr_datapath.h"
+
+#include <gr_log.h>
+#include <gr_net_types.h>
+
+#include <rte_arp.h>
+#include <rte_byteorder.h>
+#include <rte_errno.h>
+#include <rte_ether.h>
+#include <rte_graph_worker.h>
+#include <rte_icmp.h>
+#include <rte_ip.h>
+#include <rte_mbuf.h>
+
+void trace_packet(const char *node, const char *iface, const struct rte_mbuf *m) {
+	char buf[BUFSIZ], src[64], dst[64];
+	const struct rte_ether_hdr *eth;
+	uint16_t ether_type;
+	size_t offset = 0;
+	ssize_t n = 0;
+
+	eth = rte_pktmbuf_mtod_offset(m, const struct rte_ether_hdr *, offset);
+	offset += sizeof(*eth);
+	ether_type = rte_be_to_cpu_16(eth->ether_type);
+
+	n += snprintf(
+		buf + n,
+		sizeof(buf) - n,
+		ETH_ADDR_FMT " > " ETH_ADDR_FMT,
+		ETH_BYTES_SPLIT(eth->src_addr.addr_bytes),
+		ETH_BYTES_SPLIT(eth->dst_addr.addr_bytes)
+	);
+
+	if (ether_type == RTE_ETHER_TYPE_VLAN) {
+		const struct rte_vlan_hdr *vlan;
+		uint16_t vlan_id;
+
+		vlan = rte_pktmbuf_mtod_offset(m, const struct rte_vlan_hdr *, offset);
+		offset += sizeof(*vlan);
+		vlan_id = rte_be_to_cpu_16(vlan->vlan_tci) & 0xfff;
+		ether_type = rte_be_to_cpu_16(vlan->eth_proto);
```

Comment:
No QinQ support ?

Reply:
not yet.

Reply:
It seems stacking vlan interfaces is properly checked in the cli, so we should be good.
So ok for me.